src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -299,33 +334,220 @@ private BoundExpression RewriteStringConcatenationOneExpr(BoundExpression lowere
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_String__Concat_ReadOnlySpanReadOnlySpan, out MethodSymbol stringConcatSpans, isOptional: true) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_String__op_Implicit_ToReadOnlySpanOfChar, out MethodSymbol stringImplicitConversionToReadOnlySpan, isOptional: true) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_ReadOnlySpan_T__ctor_Reference, out MethodSymbol readOnlySpanCtorRefParamGeneric, isOptional: true))
+            {
+                // One of args is a char. We can use span-based concatenation, which takes a bit more IL, but avoids allocation of a string from ToString call.
+                // And since implicit conversion from string to span is a JIT intrinsic, the resulting code ends up being faster than the one generated from the "naive" case with ToString call
+                var readOnlySpanOfChar = readOnlySpanCtorRefParamGeneric.ContainingType.Construct(_compilation.GetSpecialType(SpecialType.System_Char));
+                var readOnlySpanCtorRefParamChar = readOnlySpanCtorRefParamGeneric.AsMember(readOnlySpanOfChar);
+
+                var locals = ArrayBuilder<LocalSymbol>.GetInstance(capacity: 1);
+
+                var wrappedLeft = WrapIntoReadOnlySpanOfCharIfNecessary(loweredLeft, stringImplicitConversionToReadOnlySpan, readOnlySpanCtorRefParamChar, out var leftTempLocal);
+                var wrappedRight = WrapIntoReadOnlySpanOfCharIfNecessary(loweredRight, stringImplicitConversionToReadOnlySpan, readOnlySpanCtorRefParamChar, out var rightTempLocal);
+
+                // We expect that only 1 of the operands is of char type here since `<char> + <char>` leads to operands being implicitly converted to ints and added as numbers.
+                // As a consequence of that we can produce at max 1 additional temp local variable due to the fact that only some char operands require a temp
+                // while a string operand is just always wrapped into an implicit conversion call
+                Debug.Assert(leftIsChar != rightIsChar);
+                Debug.Assert(leftTempLocal is null || rightTempLocal is null);
+
+                locals.AddIfNotNull(leftTempLocal);
+                locals.AddIfNotNull(rightTempLocal);
+
+                var concatCall = BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, stringConcatSpans, wrappedLeft, wrappedRight);
+
+                var oldSyntax = _factory.Syntax;
+                _factory.Syntax = syntax;
+
+                var sequence = _factory.Sequence(
+                    locals.ToImmutableAndFree(),
+                    ImmutableArray<BoundExpression>.Empty,
+                    concatCall);
+
+                _factory.Syntax = oldSyntax;
+                return sequence;
+            }
+
+            Debug.Assert(!leftType.IsReadOnlySpanChar() && !rightType.IsReadOnlySpanChar());
```

Comment:
>Debug.Assert(!leftType.IsReadOnlySpanChar() && !rightType.IsReadOnlySpanChar()); [](http://example.com/codeflow?start=12&length=80)

I guess it is fine to keep the assert, but it looks redundant, given the asserts at the beginning of the function